solidity/contracts/bridge/Bridge.sol

```diff
@@ -307,34 +361,410 @@ contract Bridge is Ownable {
+    /// @param mainUtxo Data of the wallet's main UTXO. If no main UTXO
+    ///        exists for given wallet, this parameter is ignored.
+    /// @dev Requirements:
+    ///      - `sweepTx` components must match the expected structure. See
+    ///        `BitcoinTx.Info` docs for reference. Their values must exactly
+    ///        correspond to appropriate Bitcoin transaction fields to produce
+    ///        a provable transaction hash.
+    ///      - The `sweepTx` should represent a Bitcoin transaction with 1..n
+    ///        inputs referring to 0..n P2(W)SH revealed unswept deposits UTXOs
+    ///        and 0..1 main P2(W)PKH UTXO. That transaction must have only
+    ///        one P2(W)PKH output containing a 20-byte wallet public key hash.
+    ///      - `sweepProof` components must match the expected structure. See
+    ///        `BitcoinTx.Proof` docs for reference. The `bitcoinHeaders`
+    ///        field must contain a valid number of block headers, not less
+    ///        than the `txProofDifficultyFactor` contract constant.
+    ///      - `mainUtxo` components must point to the recent main UTXO
+    ///        of the given wallet. If there is no main UTXO, this parameter
+    ///        is ignored.
+    function submitSweepProof(
         BitcoinTx.Info calldata sweepTx,
-        bytes memory merkleProof,
-        uint256 txIndexInBlock,
-        bytes memory bitcoinHeaders
+        BitcoinTx.Proof calldata sweepProof,
+        BitcoinTx.UTXO calldata mainUtxo
     ) external {
-        // TODO We need to read `fundingTxHash`, `fundingOutputIndex` from
-        //      `sweepTx.inputVector`. We then hash them to obtain deposit
-        //      identifier and read DepositInfo. From DepositInfo we know what
-        //      amount was inferred during deposit reveal transaction and we
-        //      use that amount to update their Bank balance, minus fee.
-        //
-        // TODO We need to validate if the sum in the output minus the
-        //      amount from the previous wallet balance input minus fees is
-        //      equal to the amount by which Bank balances were increased.
-        //
-        // TODO We need to validate `sweepTx.outputVector` to see if the balance
-        //      was not transferred away from the wallet before increasing
-        //      balances in the bank.
-        //
-        // TODO Delete deposit from unswept mapping or mark it as swept
-        //      depending on the gas costs. Alternatively, do not allow to
-        //      use the same TX input vector twice. Sweep should be provable
-        //      only one time.
+        // The actual transaction proof is performed here. After that point, we
+        // can assume the transaction happened on Bitcoin chain and has
+        // a sufficient number of confirmations as determined by
+        // `txProofDifficultyFactor` constant.
+        bytes32 sweepTxHash = validateSweepTxProof(sweepTx, sweepProof);
```

Comment:
I think it's not that simple. To validate the main UTXO, you need the `walletPubKeyHash`. To get `walletPubKeyHash` you need to `processSweepTxOutput`. To `processSweepTxOutput` you need to have the transaction data validated and proven.

Reply:
My understanding is that `processSweepTxOutput` is cheap to execute. If so can we do:

```
(bytes20 walletPubKeyHash, uint64 sweepTxOutputValue) = processSweepTxOutput(sweepTx.outputVector);

BitcoinTx.UTXO memory resolvedMainUtxo = BitcoinTx.UTXO(bytes32(0), 0, 0);
bytes32 mainUtxoHash = mainUtxos[walletPubKeyHash];
if (mainUtxoHash != bytes32(0)) {
  // (...)
}

bytes32 sweepTxHash = validateSweepTxProof(sweepTx, sweepProof);
```
?

In other words, we'd execute `validateSweepTxProof` after the `require` from line 410, comparing hashed `mainUtxo` parameter data against the one resolved from `mainUtxos` mapping.

Then, if the maintainer got frontrunned the TX will revert *before* validating the SPV proof, so this will be cheaper.

If this change is not a simple one because tests will start failing, let's just leave a TODO at the beginning of `submitSweepProof` to fail early if the function call gets frontrunned. Another option is that we can provide `walletPubKeyHash` as a parameter, resolve main UTXO from the mapping, execute the require, and validate the provided `walletPubKeyHash` later.

Reply:
We synced with @lukasz-zimnoch out-of-band and `validateSweepTxProof` is doing a static validation of inputs as well. Just changing the order is not going to work. We will address it by just adding a TODO about frontrunning and revisit in the future.

Reply:
See https://github.com/keep-network/tbtc-v2/pull/106/commits/19025d19f7301719469ba6f6ef49e3065c9a008c